### PR TITLE
Improve S3 request signing performance

### PR DESCRIPTION
**Pull request by Woellchen**
# What problem are we solving?
S3 uploading is quite CPU intense.

# How are we solving the problem?
This change is caching HMAC hashers for repeated use in subsequent requests and chunks, so they don't have to be initialized from scratch every time.
On my local computer this gives me ~5-6 times faster signature calculation and ~5-6.5% more throughput in S3 requests. The smaller the payload the better the throughput gets.

# How is the PR tested?
Existing unit tests + warp benchmarking tool.

# Checks
- [ ] I have added unit tests if possible.
- [ ] I will add related wiki document changes and link to this PR after merging.

### CODE REVIEW

File: weed/s3api/auth_signature_v4.go [hunk 1]
```diff
@@ -458,6 +463,38 @@ func (iam *IdentityAccessManagement) doesPresignedSignatureMatch(hashedPayload s
...
+	date := t.Format(yyyymmdd)
+	hashID := "AWS4" + secretKey + "/" + date + "/" + region + "/" + service + "/" + "aws4_request"
+
+	iam.hashMu.RLock()
+	pool, ok := iam.hashes[hashID]
+	iam.hashMu.RUnlock()
+
+	if !ok {
+		iam.hashMu.Lock()
+		if pool, ok = iam.hashes[hashID]; !ok {
```
Comment (chrislusf) [inline weed/s3api/auth_signature_v4.go]: how to free the map entries?

Comment (Woellchen) [reply weed/s3api/auth_signature_v4.go]: You are right I totally forgot about it. I will add some cleanup logic as well

Comment (Woellchen) [reply weed/s3api/auth_signature_v4.go]: I have added logic to clean old entries from the map

File: weed/s3api/auth_signature_v4.go [hunk 2]
```diff
@@ -458,6 +464,69 @@ func (iam *IdentityAccessManagement) doesPresignedSignatureMatch(hashedPayload s
...
+	iam.hashMu.RUnlock()
+
+	if !ok {
+		iam.hashMu.Lock()
+		defer iam.hashMu.Unlock()
+		pool, ok = iam.hashes[hashID]
+	}
+
+	if ok {
+		atomic.AddInt64(iam.hashCounters[hashID], 1)
```
Comment (Dimonyga) [inline weed/s3api/auth_signature_v4.go]: atomic.StoreInt64(iam.hashCounters[hashID], 1) ?
and maybe Int32 will be enough?

Comment (Woellchen) [reply weed/s3api/auth_signature_v4.go]: Thanks, makes sense, I adjusted it :)

**Pull request merged**